feat(core): Redesign the afterRender & afterNextRender phases API

mmalerba opened the pull request:
Previously `afterRender` and `afterNextRender` allowed the user to pass
a phase to run the callback in as part of the `AfterRenderOptions`. This
worked, but made it cumbersome to coordinate work between phases.

```ts
let size: DOMRect|null = null;

afterRender(() => {
  size = nativeEl.getBoundingClientRect();
}, {phase: AfterRenderPhase.EarlyRead});

afterRender(() => {
  otherNativeEl.style.width = size!.width + 'px';
}, {phase: AfterRenderPhase.Write});
```

This PR replaces the old phases API with a new one that allows passing a
callback per phase in a single `afterRender` / `afterNextRender` call.
The return value of each phase's callback is passed to the next phase
callback that was part of the same `afterRender` call.

```ts
afterRender({
  earlyRead: () => nativeEl.getBoundingClientRect(),
  write: (rect) => {
    otherNativeEl.style.width = rect.width + 'px';
  }
});
```

This API also retains the ability to pass a single callback, which will
be run in the `mixedReadWrite` phase.

```ts
afterRender(() => {
  // read some stuff ...
  // write some stuff ...
});
```

mmalerba committed:
feat(core): Redesign the `afterRender` & `afterNextRender` phases API

Previously `afterRender` and `afterNextRender` allowed the user to pass
a phase to run the callback in as part of the `AfterRenderOptions`. This
worked, but made it cumbersome to coordinate work between phases.

```ts
let size: DOMRect|null = null;

afterRender(() => {
  size = nativeEl.getBoundingClientRect();
}, {phase: AfterRenderPhase.EarlyRead});

afterRender(() => {
  otherNativeEl.style.width = size!.width + 'px';
}, {phase: AfterRenderPhase.Write});
```

This PR replaces the old phases API with a new one that allows passing a
callback per phase in a single `afterRender` / `afterNextRender` call.
The return value of each phase's callback is passed to the subsequent
callbacks that were part of that `afterRender` call.

```ts
afterRender({
  earlyRead: () => nativeEl.getBoundingClientRect(),
  write: (rect) => {
    otherNativeEl.style.width = rect.width + 'px';
  }
});
```

This API also retains the ability to pass a single callback, which will
be run in the `mixedReadWrite` phase.

```ts
afterRender(() => {
  // read some stuff ...
  // write some stuff ...
});
```

atscott commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -174,20 +263,74 @@ export function internalAfterNextRender(
+ *   constructor() {
+ *     afterRender(() => {
+ *       console.log('content height: ' + this.contentRef.nativeElement.scrollHeight);
+ *     }, {phase: AfterRenderPhase.Read});
```
I think this example needs to be updated. There's another in `afterRender` as well that uses `AfterRenderPhase`

atscott commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -238,37 +386,119 @@ export function afterRender(callback: VoidFunction, options?: AfterRenderOptions
+
   const destroy = () => {
-    callbackHandler.unregister(instance);
+    for (const instance of instances) {
```
Can we reorder some things here so we aren't referring to variables that aren't yet defined? 

 - move `unregisterFn` before this destroy method
 - move this destroy variable below the `instances` variable

mmalerba replied:
I can only get the `unregisterFn` by calling `onDestroy` with the `destroy` function. I can split the declaration / definition for `unregisterFn` if you prefer to see it declared first.

atscott commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -317,30 +554,72 @@ export function afterNextRender(
+  phase: AfterRenderPhase,
+  result: unknown,
+) {
+  switch (phase) {
```
I know this is kind of designing inside the PR but would it be simpler if the results were just merged with the previous result (`chainedResult = {...chainedResult, ...result};`)? The API would be a bit simpler in this case. You'd be able to consistently access results in a phase rather than needing to change where you grab them from. I guess it requires the return type to be an object with k/v pairs but maybe that's good? Kind of hard to understand what `results.write` would mean when the return value is a primitive. Much easier if it were `results.divSize`

JoostK replied:
Thanks for the examples! I do also think I prefer option 3; you can easily achieve named properties as in 2 and if you were to need it across multiple phases, spreading the prior result into the intermediate phase is easily done manually. By not needing to create an object the baisc one-phase passthrough stays simple.

alxhub replied:
How does option 3 work if I write `earlyRead` and `read` but not a `write` phase?

alxhub replied:
Option 3 feels the most ergonomic to me.

atscott replied:
👍 Option 3 SGTM as well

devknoll replied:
Option 3 is how I'd expect it to work too (i.e. as a pipeline) 👍 

I spent a little time thinking about how that can be typed, so I'd be happy to brainstorm it with you, if you want 🙂

JoostK commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -238,37 +386,119 @@ export function afterRender(callback: VoidFunction, options?: AfterRenderOptions
-  callbackHandler.register(instance);
+  const chainedResult: AfterRenderPhaseResults<any, any, any> = {};
+
+  const phases = [
+    ...(spec.earlyRead ? [[AfterRenderPhase.EarlyRead, spec.earlyRead] as const] : []),
+    ...(spec.write ? [[AfterRenderPhase.Write, spec.write] as const] : []),
+    ...(spec.mixedReadWrite
+      ? [[AfterRenderPhase.MixedReadWrite, spec.mixedReadWrite] as const]
+      : []),
+    ...(spec.read ? [[AfterRenderPhase.Read, spec.read] as const] : []),
+  ];
+
+  const instances = phases.map(([phase, callback]) => {
+    const instance = runInInjectionContext(
+      injector,
+      () => new AfterRenderCallback(phase, chainedResult, callback),
+    );
+    callbackHandler.register(instance);
+    return instance;
+  });
+
```
Avoiding the intermediate array shuffling makes things easier to follow to me:

```ts
const createCallback = (phase: AfterRenderPhase, callback: VoidFunction) {
  const instance = runInInjectionContext(
    injector,
    () => new AfterRenderCallback(phase, chainedResult, callback),
  );
  callbackHandler.register(instance);
  return instance;
};
const instances: AfterRenderCallback[] = [];
spec.earlyRead && instances.push(createCallback(AfterRenderPhase.EarlyRead, spec.earlyRead);
spec.write && instances.push(createCallback(AfterRenderPhase.Write, spec.write));
spec.mixedReadWrite && instances.push(createCallback(AfterRenderPhase.MixedReadWrite, spec.mixedReadWrite));
spec.read && instances.push(createCallback(AfterRenderPhase.Read, spec.read));
```

JoostK reviewed (commented):
I love this API 🥳

devknoll commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -20,6 +20,109 @@ import {NgZone} from '../zone/ng_zone';
+   * <div class="alert is-critical">
+   *
+   * Using this value can **significantly** degrade performance.
+   * Instead, prefer refactoring into multiple callbacks using a more specific phase.
```
> Instead, prefer dividing work into the appropriate phase callback.

devknoll commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -20,6 +20,109 @@ import {NgZone} from '../zone/ng_zone';
+ */
+export type AfterRenderPhaseSpec<E, W, M> = {
+  /**
+   * Use this phase to **read** from the DOM before a subsequent `write` callback, for example to
```
Since the phases are no longer values, it might make sense to reword the documentation here a bit more. E.g. instead of:

> **Never** use this phase to read from the DOM.

We could say:

> **Never** read from the DOM in this phase callback.

devknoll commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -174,20 +263,74 @@ export function internalAfterNextRender(
  * - in the order it was registered
  * - once per render
  * - on browser platforms only
+ * - diring the `mixedReadWrite` phase
```
> during

devknoll reviewed (commented):
Love this change. Thanks for putting it together!

mmalerba commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -317,30 +554,72 @@ export function afterNextRender(
   const destroy = () => {
-    callbackHandler.unregister(instance);
+    for (const instance of instances) {
+      if (instance) {
```
I think it was possible at some point while I was mid-implementation, but no longer

mmalerba commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -20,6 +20,97 @@ import {NgZone} from '../zone/ng_zone';
+ *
+ * @developerPreview
+ */
+export type AfterRenderPhaseFnArgs<T extends any[]> =
```
Which of these types should be part of the public API? `Head`, `Tail`, and `FirstNonNever` all feel like clearly internal helper things. I could see not making `AfterRenderPhaseFnArgs` and `AfterRenderPhaseFn` public, though they might be useful to someone trying to wrap `afterRender` in their own function (I saw somebody doing that in g3 already)

atscott commented on packages/core/src/render3/after_render_hooks.ts:
```diff
@@ -317,27 +548,50 @@ export function afterNextRender(
-
-  callbackHandler.register(instance);
+
+  const registerCallback = (
+    phase: AfterRenderPhase,
+    callback: AfterRenderPhaseFn<[] | [unknown], unknown> | undefined,
+  ) => {
+    if (!callback) {
+      return;
+    }
+    const instance = runInInjectionContext(
+      injector,
+      () =>
+        new AfterRenderCallback(phase, pipelinedArgs, (...args: [unknown]) => {
+          destroy();
+          callback(...args);
+        }),
+    );
+    callbackHandler.register(instance);
+    instances.push(instance);
+  };
+
+  registerCallback(AfterRenderPhase.EarlyRead, spec.earlyRead);
+  registerCallback(AfterRenderPhase.Write, spec.write);
+  registerCallback(AfterRenderPhase.MixedReadWrite, spec.mixedReadWrite);
+  registerCallback(AfterRenderPhase.Read, spec.read);
```
Is there something we can share here between `afterRender` and `afterNextRender`? It looks like the only difference in this block is that it calls `destroy` the first time the callback executes. Maybe we can extract this to a function and have that be the option passed in?

```
 new AfterRenderCallback(phase, pipelinedArgs, (...args: [unknown]) => {
if (executeOnce) destroy();
          callback(...args);
        }),
```

atscott reviewed (approved):
Reviewed-for: public-api

atscott commented:
Only question I might have would be whether we want to have a migration, which I think should be relatively straightforward.

mmalerba committed:
fix(core): Add back phase flag option as a deprecated API

Adds back the ability to set the phase of an `afterRender` /
`afterNextRender` callback using the `phase` option. However, this API
is now deprecated, and the phase should instead be specified by passing
a spec object rather than a callback function.

mmalerba committed:
feat(core): Add a schematic to migrate afterRender phase flag

Adds an `ng update` migration to move users from using the phase flag
with `afterRender` / `afterNextRender` to passing a spec object instead.

thePunderWoman reviewed (approved):
reviewed-for: public-api

alxhub commented:
This PR was merged into the repository by commit ea3c80205653af109c688a4d4808143b34591d54.

spock123 commented:
Thank you guys, this is a much much more elegant API.
